File: app/lib/ta/engine.ts
// Technical engine: ATR + support/resistance clustering + stop/target.
// Ported 1:1 from ta-worker/app/engine.py.
import { pivotHighs } from './factors'

const STOP_ATR_MULT = 1.0
const TARGET_ATR_MULT = 1.0
const NO_SUPPORT_ATR_MULT = 2.0
const RR_TARGET = 2.0

const mean = (xs: readonly number[]): number => xs.reduce((a, b) => a + b, 0) / xs.length

/** Average True Range — simple mean of true ranges over `period`. */
export function atr(
  highs: readonly number[],
  lows: readonly number[],
  closes: readonly number[],
  period = 14,
): number {
  const n = closes.length
  if (n < 2) return 0.0
  const trs: number[] = []
  for (let i = 1; i < n; i++) {
    trs.push(Math.max(
      highs[i] - lows[i],
      Math.abs(highs[i] - closes[i - 1]),
      Math.abs(lows[i] - closes[i - 1]),
    ))
  }
  const window = trs.length >= period ? trs.slice(-period) : trs
  return window.length ? mean(window) : 0.0
}

export interface StopTarget { stop: number; target: number; risk: number; reward: number }

/** Stop = nearest support below price padded by ATR; target = nearest resistance above, padded. */
export function stopAndTarget(
  price: number,
  atrValue: number,
  supports: readonly number[],
  resistances: readonly number[],
): StopTarget {
  const below = supports.filter(s => s < price).sort((a, b) => b - a)
  const above = resistances.filter(r => r > price).sort((a, b) => a - b)
  const stop = below.length
    ? below[0] - STOP_ATR_MULT * atrValue
    : price - NO_SUPPORT_ATR_MULT * atrValue
  const risk = price - stop
  const target = above.length
    ? above[0] + TARGET_ATR_MULT * atrValue
    : price + RR_TARGET * risk
  return { stop, target, risk, reward: target - price }
}

/** Local minima: index i where lows[i] equals the min of the ±w window. */
function pivotLows(lows: readonly number[], w: number): number[] {
  const out: number[] = []
  const n = lows.length
  for (let i = w; i < n - w; i++) {
    let m = Infinity
    for (let k = i - w; k <= i + w; k++) if (lows[k] < m) m = lows[k]
    if (lows[i] === m) out.push(lows[i])
  }
  return out
}

/** Collapse nearby levels into cluster means (band = mult * atr). */
export function cluster(levels: readonly number[], atrValue: number, mult = 0.75): number[] {
  if (!levels.length) return []
  const band = Math.max(mult * atrValue, 1e-9)
  const ordered = [...levels].sort((a, b) => a - b)
  const clusters: number[][] = [[ordered[0]]]
  for (const lv of ordered.slice(1)) {
    const head = clusters[clusters.length - 1][0]
    if (lv - head <= band) clusters[clusters.length - 1].push(lv)
    else clusters.push([lv])
  }
  return clusters.map(mean)
}

export interface Levels { resistances: number[]; supports: number[] }

/** Cluster pivot highs above price (resistance) and pivot lows below price (support). */
export function supportResistanceLevels(
  highs: readonly number[],
  lows: readonly number[],
  _closes: readonly number[],
  atrValue: number,
  price: number,
  w = 5,
): Levels {
  const resRaw = pivotHighs(highs, w).filter(([, lv]) => lv > price).map(([, lv]) => lv)
  const supRaw = pivotLows(lows, w).filter(lv => lv < price)
  return { resistances: cluster(resRaw, atrValue), supports: cluster(supRaw, atrValue) }
}

export { pivotLows as _pivotLows }

File: ta-worker/app/engine.py
"""Technical engine: ATR + stop/target derivation.
Clustering + vol-scaled stop/target adapted from Portfolio
backend/app/services/risk_reward_service.py."""
from __future__ import annotations

import numpy as np

from app.factors import pivot_highs

STOP_ATR_MULT = 1.0
TARGET_ATR_MULT = 1.0
NO_SUPPORT_ATR_MULT = 2.0
RR_TARGET = 2.0


def atr(
    highs: list[float], lows: list[float], closes: list[float], period: int = 14
) -> float:
    """Average True Range (simple mean of true ranges over `period`)."""
    n = len(closes)
    if n < 2:
        return 0.0
    trs = []
    for i in range(1, n):
        tr = max(
            highs[i] - lows[i],
            abs(highs[i] - closes[i - 1]),
            abs(lows[i] - closes[i - 1]),
        )
        trs.append(tr)
    window = trs[-period:] if len(trs) >= period else trs
    return float(np.mean(window)) if window else 0.0


def stop_and_target(
    price: float, atr_value: float, supports: list[float], resistances: list[float]
) -> dict:
    """Stop = nearest support below price padded by ATR; target = nearest
    resistance above price padded by ATR; both have fallbacks."""
    below = sorted([s for s in supports if s < price], reverse=True)
    above = sorted([r for r in resistances if r > price])

    stop = (
        below[0] - STOP_ATR_MULT * atr_value
        if below
        else price - NO_SUPPORT_ATR_MULT * atr_value
    )
    risk = price - stop
    target = (
        above[0] + TARGET_ATR_MULT * atr_value if above else price + RR_TARGET * risk
    )
    reward = target - price
    return {"stop": stop, "target": target, "risk": risk, "reward": reward}


def _pivot_lows(lows: list[float], w: int) -> list[float]:
    """Local minima: index i where lows[i] == min of the +/- w window.

    Assumes `lows` are raw stored OHLC values compared by exact value;
    derived/resampled lows would need a tolerance (exact float equality
    is unreliable for derived values).
    """
    out = []
    n = len(lows)
    for i in range(w, n - w):
        if lows[i] == min(lows[i - w : i + w + 1]):
            out.append(float(lows[i]))
    return out


def _cluster(levels: list[float], atr: float, mult: float = 0.75) -> list[float]:
    """Collapse nearby levels into cluster means (band = mult * atr)."""
    if not levels:
        return []
    band = max(mult * atr, 1e-9)
    ordered = sorted(levels)
    clusters: list[list[float]] = [[ordered[0]]]
    for lv in ordered[1:]:
        if lv - clusters[-1][0] <= band:
            clusters[-1].append(lv)
        else:
            clusters.append([lv])
    return [float(np.mean(c)) for c in clusters]


def support_resistance_levels(
    highs: list[float],
    lows: list[float],
    closes: list[float],
    atr: float,
    price: float,
    w: int = 5,
) -> dict:
    """Cluster pivot highs above price (resistance) and pivot lows below price (support)."""
    res_raw = [lv for _, lv in pivot_highs(highs, w=w) if lv > price]
    sup_raw = [lv for lv in _pivot_lows(lows, w=w) if lv < price]
    return {"resistances": _cluster(res_raw, atr), "supports": _cluster(sup_raw, atr)}
